src/number_theory/diophantine_approximation.lean

```diff
@@ -265,3 +284,317 @@ begin
   rw [H, (by push_cast : (1 : ℝ) / q.denom ^ 2 = (1 / q.denom ^ 2 : ℚ))],
   norm_cast,
 end
+
+/-!
+### Legendre's Theorem on Rational Approximation
+
+We prove **Legendre's Theorem** on rational approximation: If $\xi$ is a real number and
+$x/y$ is a rational number such that $|\xi - x/y| < 1/(2y^2)$,
+then $x/y$ is a convergent of the continued fraction expansion of $\xi$.
+
+The proof is by induction. However, the induction proof does not work with the
+statement as given, since the assumption is too weak to imply the corresponding
+statement for the application of the induction hypothesis. This can be remedied
+by making the statement slightly stronger. Namely, we assume that $|\xi - x/y| < 1/(y(2y-1))$
+when $y \ge 2$ and $-\frac{1}{2} < \xi - x < 1$ when $y = 1$.
+-/
+
+section convergent
+
+namespace real
+
+open int
+
+/-!
+### Convergents: definition and API lemmas
+-/
+
+/-- We give a direct recursive definition of the convergents of the continued fraction
+expansion of a real number `ξ`. The main reason for that is that we want to have the
+convergents as rational numbers; the versions
+`(generalized_continued_fraction.of ξ).convergents` and
+`(generalized_continued_fraction.of ξ).convergents'` always give something of the
+same type as `ξ`. We can then also use dot notation `ξ.convergent n`.
+Another minor reason is that this demonstrates that the proof
+of Legendre's theorem does not need anything beyond this definition.
+We provide a proof that this definition agrees with the other one;
+see `real.continued_fraction_convergent_eq_convergent`.
+(Note that we use the fact that `1/0 = 0` here to make it work for rational `ξ`.) -/
+noncomputable def convergent : ℝ → ℕ → ℚ
+| ξ 0 := ⌊ξ⌋
+| ξ (n + 1) := ⌊ξ⌋ + 1 / convergent (1 / fract ξ) n
```

Comment:
`a⁻¹` is simp normal form for `1 / a`
```suggestion
| ξ (n + 1) := ⌊ξ⌋ + (convergent (fract ξ)⁻¹ n)⁻¹
```

Reply:
OK; I guess this is because `x⁻¹` is the simp normal form of `1 / x`. (I'm not sure I like this decision, but I guess it's better to follow it...)

Reply:
Does it make more sense if I tell you that `-a` is simp normal form for `0 - a`? 😁

Reply:
I figured that out. But somehow, it still feels different...